Update slicing/rendering docs (#424)
napari/docs: docs/guides/rendering.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.10.3
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

(rendering)=

# Rendering

This document explains how napari produces a 2- or 3-dimensional render in the canvas from layers' n-dimensional array-like data.

> The intended audience is someone who wants to understand napari's rendering pipeline to help optimize its performance for their usage.

Reviewer: ```suggestion
The intended audience is someone who wants to understand napari's rendering pipeline to help optimize its performance for their usage,
```

> Or someone that wants to contribute to and help improve the clarity or performance of napari's rendering pipeline itself.

Reviewer: ```suggestion
or someone who wants to contribute to and help improve the clarity or performance of napari's rendering pipeline itself.
```

## Overview

At a high level, rendering in napari is simple.

1. Viewing: {attr}`ViewerModel.dims<napari.components.ViewerModel.dims>` defines which 2D or 3D region is currently being viewed.
2. Slicing: [`Layer._slice_dims`](https://github.com/napari/napari/blob/b3a8dd22895c913d8183735f52b9d1d71c963d7f/napari/layers/base/base.py#L1184) loads the corresponding 2D or 3D region of the layer's ND data into RAM.
3. Drawing: [`VispyBaseLayer._on_data_change`](https://github.com/napari/napari/blob/b3a8dd22895c913d8183735f52b9d1d71c963d7f/napari/_vispy/layers/base.py#L126) pushes the 2D or 3D sliced data from RAM to VRAM to be drawn on screen.

But as the details of this document reveal, rendering in napari is, in fact, very complicated.

Consider some of the more important reasons for this.

- Multiple layers can have different extents with different transforms.
- Different layer types (e.g. Images vs. Points) handle slicing differently.
- Layer data can be large or slow to load into RAM.

> - Sliced data in VRAM may exceed the maximum texture size.

Reviewer: Can we expand just with a few more words here for newbies like me sorry? I would love to see something like "maximum texture size of the graphics card" or "maximum texture size of vispy" or idk, as you can tell I don't know who's maximum texture size this is 😆

Author: Incoming.

in addition to the `step` that should be taken in each dimension as its corresponding slider position changes.

In our running example, the range

```{code-cell} python
viewer.dims.range
```

is solely determined by the shape of the data

```{code-cell} python
viewer.layers[0].data.shape
```

because the layers have the same shape and identity transforms.

### Point and selective slicing

{attr}`Dims.point<napari.components.Dims.point>` describes the coordinates of the current slice plane in that same world coordinate system.

In our running example, the default 2D view defines {attr}`Dims.point<napari.components.Dims.point>` to be

```{code-cell} python
viewer.dims.point
```

which represents the mid-point through all three dimensions.
As the last two dimensions are visualized in the canvas, this represents the 2D plane through the middle of the first dimension.

Only the dimensions in {attr}`Dims.not_displayed<napari.components.Dims.not_displayed>` have meaningful values in
{attr}`Dims.point<napari.components.Dims.point>` because all data in displayed dimensions is retained in a slice,

> even though those data may not be visible in the canvas due to the current camera parameter values.

Reviewer: so this is to say that even if I'm zoomed in and can only see a part of the current 2D slice, all the data in the slice is still loaded, right?

Reviewer 2: Yes, unless it's multiscale/tiled.

```{code-cell} python
import numpy as np
from napari.layers import Image

mean_data = np.mean(viewer.layers[0].data, axis=0)
viewer.layers[0] = Image(mean_data, colormap=viewer.layers[0].colormap)
world_dims = np.asarray(viewer.dims.order)
layer0_dims = viewer.layers[0]._world_to_layer_dims(
  world_dims=world_dims, ndim_world=3)
layer1_dims = viewer.layers[1]._world_to_layer_dims(
  world_dims=world_dims, ndim_world=3)
print(f'{layer0_dims=}')
print(f'{layer1_dims=}')
```

where `Layer._world_to_layer_dims` is a private method that is called as a part of slicing.

For simple cases like the above, this right-alignment approach tends to work well.

But for more complex cases, it quickly runs into problems.
For example, changing the dimensions that are displayed in the canvas (by changing {attr}`Dims.order<napari.components.Dims.order>`)
causes the mapping to change (see [issue #3882](https://github.com/napari/napari/issues/3882)).

We would love to fix these problems.
There are few related issues and conversations, but maybe the best way to track our progress is to follow [issue #5949](https://github.com/napari/napari/issues/5949)

> which aims enrich napari's handling of dimensions in general.

Reviewer: ```suggestion
which aims to enrich napari's handling of dimensions in general.
```

### Loading layer data

Once we have the slice indices into a layer's data, we need to load the corresponding region of data into RAM.

### Loading array-like image data

{class}`Image<napari.layers.Image>` layer data does not have a single specific type (e.g. numpy's `ndarray`).
Instead it must only have the attributes and methods defined in [`LayerDataProtocol`](https://github.com/napari/napari/blob/eab7661459e70479c7c7d587a36463f3b099b64a/napari/layers/_data_protocols.py#L51).

Numpy's `ndarray` is compatible with this protocol, but so are array types from other packages like [Dask](https://docs.dask.org/en/latest/array.html), [Zarr](https://zarr.readthedocs.io/en/stable/_autoapi/zarr.core.Array.html), and more.
This flexibility allows you refer to image data that does not fit in memory or still needs to be lazily computed, without complicating napari's core implementation at all.

However, it also means that simply reading image data may be slow because the data must be read from disk, downloaded across a network, or calculated from a compute graph.
This means array accesses can take an arbitrary long time to complete.

### Loading multi-scale image data

> `Image` layers also support multi-scale image data, where multiple resolutions of the same image content are stored.

Reviewer: True also for labels right? Maybe worth mentioning once somewhere?

Reviewer 3: Yes, excellent points. With the caveat that one can't paint into multiscale labels.

Author: I added `Labels` here to make this clear. There are a lot of details in this document already, so I did not add the caveat because there are many many more.

Similarly to regular image data, this is supported by defining a [`MultiScaleData`](https://github.com/napari/napari/blob/eab7661459e70479c7c7d587a36463f3b099b64a/napari/layers/_multiscale_data.py#L13) protocol.
As this protocol is mostly just `Sequence[LayerDataProtocol]`, this comes with the same flexibility and arbitrary load times.

However, rendering multi-scale image data differs from regular image data because we must choose which scale or data level to load.
In order to do this, [`compute_multiscale_level`](https://github.com/napari/napari/blob/40ac1fb242d905d503aed8200099efd02ebceb95/napari/layers/utils/layer_utils.py#L532)
uses the canvas' field of view and the canvas' size in screen pixels to find the finest resolution data level that ensures that there is at least one layer data pixel per screen pixel.

> As a part of these calculates, {attr}`Layer.corner_pixels<napari.layers.Layer.corner_pixels>` is updated to store the top-left and bottom-right corner of the canvas' field of view in the data coordinates of the currently rendered level.

Reviewer: ```suggestion
As a part of these calculations, {attr}`Layer.corner_pixels<napari.layers.Layer.corner_pixels>` is updated to store the top-left and bottom-right corner of the canvas' field of view in the data coordinates of the currently rendered level.
```

The key code changes push all slicing (including synchronous slicing) through a dedicated controller [`_LayerSlicer`](https://github.com/napari/napari/blob/b3a8dd22895c913d8183735f52b9d1d71c963d7f/napari/components/_layer_slicer.py#L80),
and define all the layer-specific slicing logic in dedicated callable classes (e.g. [`_ImageSliceRequest`](https://github.com/napari/napari/blob/b3a8dd22895c913d8183735f52b9d1d71c963d7f/napari/layers/image/_slice.py#L154)).
An instance of one of these callables captures all the state needed to perform slicing,
so that it can be executed asynchronously on another thread without needing to guard competing access to that state with locks.

Unfortunately, these new additions make following the old synchronous slicing code paths more complicated.
But eventually we hope to mostly remove those complications and make both synchronous and asynchronous slicing consistent and easy enough to follow.

#### Future

The current experimental asynchronous slicing approach is limited.
While it prevents napari from blocking the main thread, fetching and rendering the data in view can still be slow.

> Most large image viewers improve on this experience with by progressively fetching and rendering chunks or tiles of data.

Reviewer: ```suggestion
Most large image viewers improve on this experience by progressively fetching and rendering chunks or tiles of data.
```

This allows some data to be presented quickly rather than waiting for everything in view,
which often results in a much better user experience when fetching the data is slow.

These efforts across all layers are generally tracked by [issue #5942](https://github.com/napari/napari/issues/5942).
Currently, the most focus is on the image layer in [issue #5561](https://github.com/napari/napari/issues/5561),
with lots of progress towards that in [PR #6043](https://github.com/napari/napari/pull/6043).

## Drawing

After the current view of a layer's data has been sliced into RAM, the data is pushed to VRAM with any associated transforms and parameters.

Each napari layer type has a corresponding vispy layer type.
For example, the [`VispyImageLayer`](https://github.com/napari/napari/blob/5e8dc098cb213c5f963524e619f223ad4fe90be8/napari/_vispy/layers/base.py#L21)
corresponds to the {class}`Image<napari.layers.Image>` layer.
Each instance of a layer also has a corresponding instance of its vispy layer.

> These correspondences can be found [`VispyCanvas.layer_to_visual`](https://github.com/napari/napari/blob/5e8dc098cb213c5f963524e619f223ad4fe90be8/napari/_vispy/canvas.py#L69).

Reviewer: ```suggestion
These correspondences can be found in [`VispyCanvas.layer_to_visual`](https://github.com/napari/napari/blob/5e8dc098cb213c5f963524e619f223ad4fe90be8/napari/_vispy/canvas.py#L69).
```

> The vispy layer instance has a reference to its corresponding layer.

Reviewer: Above you say 

> Each instance of a layer also has a corresponding instance of its vispy layer.

So does that mean that the layer instance and vispy layer instance both hold references to each other?

Reviewer 2: Yeah maybe this should be worded differently. Vispy side holds references to the models and uses events to get updates, but the models know nothing of the vispy side.

Author: I reworded

> Each instance of a layer also has a corresponding instance of its vispy layer.

to

> For each instance of a layer, there is a corresponding instance of its type's vispy layer

so that it doesn't sound like the layer has/owns the vispy layer.

> Updates to the layer's state and its current slice trigger the vispy layer are handled using [napari's event system](connect-napari-event).

Reviewer: I actually wasn't sure here...

```suggestion
Updates to the layer's state and its current slice triggered by the vispy layer are handled using [napari's event system](connect-napari-event).
```

Author: I just removed `trigger the vispy layer` here because I think that's explained underneath.

Of particular interest here is the [`Layer.events.set_data` event](layer-events), which is connected to the abstract method
[`VispyBaseLayer._on_data_change`](https://github.com/napari/napari/blob/5e8dc098cb213c5f963524e619f223ad4fe90be8/napari/_vispy/layers/base.py#L74).
This event is triggered when slicing is finished and the latest slice state can be read.

Each vispy layer type is responsible for implementing the `_on_data_change` method.

> This implementation of this method should read the updated state from the layer, then update the vispy layer appropriately.

Reviewer: ```suggestion
The implementation of this method should read the updated state from the layer, then update the vispy layer appropriately.
```

> In turn, vispy handles make the appropriate updates to VRAM and executes any programs needed to update the display on napari's canvas.

Reviewer: ```suggestion
In turn, vispy handles making the appropriate updates to VRAM and executes any programs needed to update the display on napari's canvas.
```

Author: Using `makes` instead of `handles making` instead.
